openmmtools/alchemy.py

```diff
@@ -12,413 +12,905 @@
+            OpenMM tries to select the fastest available.
+
+        Returns
+        -------
+        energy_components : dict str: simtk.unit.Quantity
+            A string label describing the role of the force associated to
+            its contribution to the potential energy.
 
         """
-        platform = openmm.Platform.getPlatformByName('Reference')
+        # Find and label all forces.
+        force_labels = cls._find_force_components(alchemical_system)
+        assert len(force_labels) <= 32, ("The alchemical system has more than 32 force groups; "
+                                         "can't compute individual force component energies.")
+
+        # Create deep copy of alchemical system.
+        system = copy.deepcopy(alchemical_system)
+
+        # Separate all forces into separate force groups.
+        for force_index, force in enumerate(system.getForces()):
+            force.setForceGroup(force_index)
+
+        # Create a Context in the given state.
         integrator = openmm.VerletIntegrator(1.0 * unit.femtoseconds)
-        context = openmm.Context(system, integrator, platform)
-        alchemical_parameters = [ parameter for parameter in context.getState(getParameters=True).getParameters().keys() if parameter.startswith(prefix) ]
+        if platform is None:
+            context = openmm.Context(system, integrator)
+        else:
+            context = openmm.Context(system, integrator, platform)
+        context.setPositions(positions)
+        alchemical_state.apply_to_context(context)
+
+        # Get energy components
+        energy_components = collections.OrderedDict()
+        for force_label, force_index in force_labels.items():
+            energy_components[force_label] = context.getState(getEnergy=True,
+                                                              groups=2**force_index).getPotentialEnergy()
+        # Clean up
         del context, integrator
-        return alchemical_parameters
 
-    def NoninteractingAlchemicalState(self):
-        """
-        Return a noninteracting alchemical state where all alchemical parmeters are set to zero.
-        """
-        kwargs = { parameter : 0 for parameter in self.alchemical_parameters }
-        return AlchemicalState(**kwargs)
+        return energy_components
 
-    def FullyInteractingAlchemicalState(self):
-        """
-        Return a noninteracting alchemical state where all alchemical parmeters are set to unity.
-        """
-        kwargs = { parameter : 1 for parameter in self.alchemical_parameters }
-        return AlchemicalState(**kwargs)
+    # -------------------------------------------------------------------------
+    # Internal usage: AlchemicalRegion
+    # -------------------------------------------------------------------------
 
     @classmethod
-    def _tabulateBonds(cls, system):
+    def _resolve_alchemical_region(cls, system, alchemical_region):
```

Comment:
Pretty much it is to ensure that all `alchemical_X` are `set`s and not `bool`s. If the passed arguments are `bool`s this construct the list of indices from the `alchemical_atoms` list.